### PR TITLE
Fixed #30123 -- Removed tuple support in DatabaseIntrospection.get_field_type().

### PR DESCRIPTION
**Pull request by pope1ni**
Ticket [#30123](https://code.djangoproject.com/ticket/30123).

Split out from #10871, targeting 3.0.

Note that this includes the commits from #10883 upon which this builds.

### CODE REVIEW

**Comment by pope1ni** [comment]
I expect tests to fail until #10871 is merged.

File: django/contrib/gis/db/backends/postgis/introspection.py [hunk 1]
```diff
@@ -15,45 +13,20 @@ class PostGISIntrospection(DatabaseIntrospection):
...
-            self.postgis_types_reverse = self.get_postgis_types()
-            self.data_types_reverse.update(self.postgis_types_reverse)
+        if not self.postgis_oid_lookup:
+            # To determine the OID integers for the PostGIS data types used in
+            # reverse lookup, query PostgreSQL's pg_type table (the integers
+            # may be different across versions). To prevent unnecessary
+            # requests upon connection initialization, the `data_types_reverse`
+            # dictionary isn't updated until introspection is performed here.
+            with self.connection.cursor() as cursor:
+                cursor.execute('SELECT oid, typname FROM pg_type WHERE typname IN (%s, %s)', ('geometry', 'geography'))
```
Comment (timgraham) [inline django/contrib/gis/db/backends/postgis/introspection.py]: Is there an advantage to putting `('geometry', 'geography')` as params? `"SELECT oid, typname FROM pg_type WHERE typname IN ('geometry', 'geography')"` works as far as I tested.

Comment (pope1ni) [reply django/contrib/gis/db/backends/postgis/introspection.py]: No. Well observed!

File: django/contrib/gis/db/backends/postgis/introspection.py [hunk 2]
```diff
@@ -15,45 +13,20 @@ class PostGISIntrospection(DatabaseIntrospection):
...
     def get_field_type(self, data_type, description):
-        if not self.postgis_types_reverse:
-            # If the PostGIS types reverse dictionary is not populated, do so
-            # now.  In order to prevent unnecessary requests upon connection
-            # initialization, the `data_types_reverse` dictionary is not updated
-            # with the PostGIS custom types until introspection is actually
-            # performed -- in other words, when this function is called.
-            self.postgis_types_reverse = self.get_postgis_types()
-            self.data_types_reverse.update(self.postgis_types_reverse)
+        if not self.postgis_oid_lookup:
```
Comment (claudep) [inline django/contrib/gis/db/backends/postgis/introspection.py]: Just as a side note, I already hoped several time that we provide a (decorator?) utility to help with this pattern of a class-level cached data (some combination between classmethod and cached_property). That would be nice to have!

**Pull request merged**